Mathlib/AlgebraicGeometry/Morphisms/UnderlyingMap.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.Topology.LocalAtTarget
+import Mathlib.AlgebraicGeometry.Morphisms.Constructors
+
+/-!
+
+## Properties on the underlying functions of morphisms of schemes.
+
+This file includes various results on properties of morphisms of schemes that come from properties
+of the underlying map, including
+
+- `Injective`
+- `Surjective`
+- `IsOpenMap`
+- `IsClosedMap`
+- `Embedding`
+- `OpenEmbedding`
+- `ClosedEmbedding`
+
+
+-/
+
+open CategoryTheory
+
+namespace AlgebraicGeometry
+
+universe u
+
+section Injective
+
+variable {X Y Z : Scheme.{u}} (f : X ⟶ Y) (g : Y ⟶ Z)
+
+instance : MorphismProperty.RespectsIso (topologically Function.Injective) :=
+  topologically_respectsIso _ (fun e ↦ e.injective) (fun _ _ hf hg ↦ hg.comp hf)
+
+instance injective_isLocalAtTarget : IsLocalAtTarget (topologically Function.Injective) := by
+  refine topologically_isLocalAtTarget _ (fun _ s h ↦ h.restrictPreimage s)
+    fun f ι U H _ hf x₁ x₂ e ↦ ?_
+  obtain ⟨i, hxi⟩ : ∃ i, f x₁ ∈ U i := by simpa using congr(f x₁ ∈ $H)
+  exact congr(($(@hf i ⟨x₁, hxi⟩ ⟨x₂, show f x₂ ∈ U i from e ▸ hxi⟩ (Subtype.ext e))).1)
+
+end Injective
+
+section Surjective
+
+variable {X Y Z : Scheme.{u}} (f : X ⟶ Y) (g : Y ⟶ Z)
+
+/-- A morphism of schemes is surjective if the underlying map is. -/
+@[mk_iff]
+class Surjective : Prop where
+  surj : Function.Surjective f.1.base
+
```

Comment:
I am wondering if we should agree on a convention, whether we write `f.val.base` or `f.1.base` or even `f.1.1`. I personally prefer the first version as I think it is the most readable one.

Reply:
I personally think spelling it out is a good thing but mathlib (or at least I) uses `.1` and `.val` interchangeably so I end up using `f.1.base` most of the time. I don't think the vague `val` gives much information that is worth +2 chars but I'm not against a library-wide replacement if you feel like it.

Reply:
At least `.val` indicates that it is a subtype where `.1` can really be anything. In any case, this is not important for this PR.